# HG changeset patch
# User Magnus Melin <mkmelin+mozilla@iki.fi>
# Date 1745770059 -7200
# Node ID 63eca98887dd4e306fcb8530f900e3ce91d1059e
# Parent  5f60e1fce4b5cf27cdd0d8b8a67e46d7a1788547
Bug 1851297 - Fix shutdownhang due to some problem in nsOutlookCompose::ComposeTheMessage. r=tobyp

This won't fix the problem with the hanging import. But should prevent the crash.

Differential Revision: https://phabricator.services.mozilla.com/D245044

diff --git a/mailnews/import/public/nsIImportService.idl b/mailnews/import/public/nsIImportService.idl
--- a/mailnews/import/public/nsIImportService.idl
+++ b/mailnews/import/public/nsIImportService.idl
@@ -2,12 +2,10 @@
 /* This Source Code Form is subject to the terms of the Mozilla Public
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
-/*
-
-  The import service.
-
+/**
+ * The import service.
  */
 
 #include "nsISupports.idl"
 
@@ -20,15 +18,19 @@ interface nsIMsgIdentity;
 interface nsIMsgCompFields;
 interface nsIMsgAttachedFile;
 interface nsIMsgEmbeddedImageData;
 
+%{ C++
+#define NS_IMPORTSERVICE_CONTRACTID "@mozilla.org/import/import-service;1"
+%}
+
 [scriptable, uuid(d0ed4c50-5997-49c9-8a6a-045f0680ed29)]
 interface nsIImportService : nsISupports {
-  nsIImportMailboxDescriptor CreateNewMailboxDescriptor();
-  nsIImportABDescriptor CreateNewABDescriptor();
-  nsIImportGeneric CreateNewGenericMail();
-  nsIImportGeneric CreateNewGenericAddressBooks();
-  void CreateRFC822Message(in nsIMsgIdentity aIdentity,
+  nsIImportMailboxDescriptor createNewMailboxDescriptor();
+  nsIImportABDescriptor createNewABDescriptor();
+  nsIImportGeneric createNewGenericMail();
+  nsIImportGeneric createNewGenericAddressBooks();
+  void createRFC822Message(in nsIMsgIdentity aIdentity,
                            in nsIMsgCompFields aMsgFields,
                            in string aBodytype,
                            in ACString aBody,
                            in boolean aCreateAsDraft,
@@ -36,13 +38,4 @@ interface nsIImportService : nsISupports
                            in Array<nsIMsgEmbeddedImageData> aEmbeddedObjects,
                            in nsIMsgSendListener aListener);
 
 };
-
-%{ C++
-#define NS_IMPORTSERVICE_CID              \
-{ /* 5df96d60-1726-11d3-a206-00a0cc26da63 */      \
-   0x5df96d60, 0x1726, 0x11d3,                   \
-   {0xa2, 0x06, 0x0, 0xa0, 0xcc, 0x26, 0xda, 0x63}}
-
-#define NS_IMPORTSERVICE_CONTRACTID "@mozilla.org/import/import-service;1"
-%}
diff --git a/mailnews/import/src/nsOutlookCompose.cpp b/mailnews/import/src/nsOutlookCompose.cpp
--- a/mailnews/import/src/nsOutlookCompose.cpp
+++ b/mailnews/import/src/nsOutlookCompose.cpp
@@ -31,8 +31,10 @@
 
 #include "nsMsgMessageFlags.h"
 #include "nsMsgLocalFolderHdrs.h"
 
+#include "mozilla/SpinEventLoopUntil.h"
+
 #define NS_MSGCOMPFIELDS_CID                         \
   { /* e64b0f51-0d7b-4e2f-8c60-3862ee8c174f */       \
     0xe64b0f51, 0x0d7b, 0x4e2f, {                    \
       0x8c, 0x60, 0x38, 0x62, 0xee, 0x8c, 0x17, 0x4f \
@@ -291,15 +293,19 @@ nsresult nsOutlookCompose::ComposeTheMes
   const char* charset = msg.GetBodyCharset();
   nsMsgI18NConvertFromUnicode(
       charset ? nsDependentCString(charset) : EmptyCString(), bodyW, bodyA);
 
-  nsCOMPtr<nsIImportService> impService(
+  nsCOMPtr<nsIMsgAccountManager> accountManager =
+      do_GetService("@mozilla.org/messenger/account-manager;1", &rv);
+  NS_ENSURE_SUCCESS(rv, rv);
+
+  nsCOMPtr<nsIImportService> importService(
       do_GetService(NS_IMPORTSERVICE_CONTRACTID, &rv));
   NS_ENSURE_SUCCESS(rv, rv);
 
-  // nsIImportService::CreateRFC822Message creates a runnable and dispatches to
+  // nsIImportService.createRFC822Message creates a runnable and dispatches to
   // the main thread.
-  rv = impService->CreateRFC822Message(
+  rv = importService->CreateRFC822Message(
       m_pIdentity,   // dummy identity
       m_pMsgFields,  // message fields
       msg.BodyIsHtml() ? "text/html" : "text/plain",
       bodyA,  // body pointer
@@ -315,12 +321,14 @@ nsresult nsOutlookCompose::ComposeTheMes
     mozilla::GetErrorName(rv, name);
     IMPORT_LOG1("*** Error, CreateAndSendMessage FAILED: %s\n", name.get());
   } else {
     // Wait for the listener to get done.
-    nsCOMPtr<nsIThread> thread(do_GetCurrentThread());
-    while (!pListen->m_done) {
-      NS_ProcessNextEvent(thread, true);
-    }
+    mozilla::SpinEventLoopUntil(
+        "nsIImportService.createRFC822Message is async"_ns, [=]() {
+          bool shutdownInProgress = false;
+          accountManager->GetShutdownInProgress(&shutdownInProgress);
+          return pListen->m_done || shutdownInProgress;
+        });
   }
 
   if (pListen->m_location) {
     pListen->m_location->Clone(pMsg);
